### ISSUE TITLE
Hi fi motor and turret model?

### ISSUE THREAD

**Issue by devangel77b** [usna-ew309/corona#7]
Good morning all,

Where would you like me to focus my efforts? I found some literature modeling DC motors with higher fidelity friction, deadzones, and directional friction included. I can make a simulation model closer to the real deal if we think that is useful.

Students can use the data to create a simple model, design controllers based on the model, then test the controller on the higher fidelity model. Thoughts?

To fit the model parameters to the actual turrets, I’ll need bi-directional data. Are there data sets with the motor turning both ways? (Step in each direction, sine wave input, etc.)

Best,

Levi

**Comment by devangel77b** [comment]
From @tracieseverson: 4.  Test their model - either test it on a higher fidelity model as Levi suggested or develop their own.  If you have time Levi - it would be nice to have a higher fidelity model they can test on.  That would give them a chance to analyze the friction and deadzone stuff.  But if it's too hard, I think we can use something similar to what you developed a few years ago (attached).

Levi - can you work on the higher fidelity model?  I can get you bi-directional data.  Right now I have a CW step and sine wave input (that I'm planning to redo this afternoon).  I'll get the other set for you shortly.

**Comment by devriesusna** [comment]
OK, I'm on it! @devangel77b @tracieseverson

**Comment by devangel77b** [comment]
From @devriesusna Hi Mike,

OK, this looks great. The only thing not captured in this model would be deadzone caused by stiction (Stribeck effect). This appears to be a relatively easy thing to add, which is a low-work way to capture the directional dependence of the motion but still be pretty similar to what their models will turn out to be.

My question for the group on implementation: What platform do we see the main simulation unfolding in? Should the end product delivered to students be a "real motor" simulink block or a function similar to "step" or "lsim" for transfer functions? I'm going to proceed with the function approach because it can be easily carried over into Simulink, but I just want to be clear so everyone is developing with the same framework in mind.

Best,

Levi

**Comment by devangel77b** [comment]
Good evening all,

End-of-day update on my end:

1) Completed simulation model of nonlinear DC motor. I added parameters for the direction-dependent dead zone and direction-dependent friction. All files pushed to github.

As you can see in the comparison between simulation and experiment, the shapes are the same, but the scale is off by orders of magnitude. This is because I made up transducer constants, moments of inertia, inductances, resistances, dead zones, friction magnitudes, etc.

**Comment by devangel77b** [comment]
TODOs:
1) Adjust model parameters to match experimental data (using sine and step data)
2) Incorporate controller function that passes PID gains as discussed in today's meeting.

QUESTION:
1) Are the units on the experimental data in radians? If so, ~30 rad is  ~5 revolutions. Tracie, was that thing spinning wildly during this test? Are these units perhaps degrees?

**Comment by devangel77b** [comment]
https://drive.google.com/drive/u/1/folders/1iJGQQEALKOLcR9fYB1UGzSCy0p8qvK3C
(model info from Feemster via Kutzer)

**Comment by devangel77b** [comment]
The units were radians.  Yes it was spinning very fast.  I wanted to at least get a few revolutions in each direction during the sine wave experiment.  The duty cycle varied from -1 to 1 so that is kind of what I expected.

  I also added the dead zone ID files to the Google Drive.  Not sure if you used those for the simulation or not.  I forgot to send an update yesterday but I created a new fold on the Google Drive for Dead Zone ID and they live there.

**Comment by devriesusna** [comment]
Model is complete. Last step is to organize how it will interface with students.

**Comment by devangel77b** [comment]
1) All the code is commented.
2) Simplified the model to eliminate motor inductance. The change didn't affect the agreement with the experimental data very much and including the electrical differential equation in the state space caused some numerical sensitivities in ode45. It speeds the code up a noticeable amount.

TODOs (those marked with * need input before I proceed):
1*) Tidy up the presentation for students to use. How should the overall framework be presented to students? We had mentioned a function like out = ClosedLoopDynamics(Kp,Ki,Kd). To incorporate with computer vision the simulation will presumably need to proceed in discrete time steps in a for or while loop, correct?
2) Simulation can be slow (~5 seconds) using ode45 when the derivative gain increases. I assume some numerical sensitivity is showing up, just need to figure out what it is doing.
3) Anything I'm missing?


Once this is done, where can I jump in and help next?

**Comment by devriesusna** [comment]
I tried to package up the nonlinear simulation similar to how I understand students will use it. I created a function SSE = sendCmdtoDcMotor(des_theta,Kp,Ki,Kd) that returns the steady-state error to the students when provided the desired pointing angle and PID gains. I tried to make it mbed-like, where they send a serial command, the mbed runs, and they get some result/data back.

There are additional variable inputs and outputs to the function that will allow students to view the entire response (position, speed, integral term) and configure initial conditions, time of simulation, etc. The documentation is relatively non-existent right now, so my apologies for that. It's on my TODO list.

Let me know if this is the direction we had in mind for how the students will interact with the nonlinear model. I can revise, start over, or add-on as we see fit.

Have a great weekend!

**Comment by devriesusna** [comment]
Good evening all,

The latest code update to the simulator (sendCmdtoDcMotor.m) allows the user to simulate the behavior in a few different operational modes by specifying a "mode" of operation input to the function. The mode can be either 'step' for a step input (applied at t=1.0 seconds) or 'closed' for a closed-loop PID controller.

I've also consolidated the sendCmdtoDcMotor function so it is self-contained (all sub-functions called live in the same m-file. This way we can provide a single m-file to the students rather than a number of files.

I've updated the documentation and tried a few different test cases, but that's about it. I'm not confident that it is mid-proof yet, so before it goes live to students you might want to put it through the paces and let me know if I've missed any bugs.

Best,

Levi
